mitodl/ocw-studio
# Added first version of site config schema + validation

#### Pre-Flight checklist


#### What are the relevant tickets?
Closes #129 

#### What's this PR do?
- Adds a schema for our site configs, and validation against that schema in places where a site config can be defined (django admin, localdev override)
- Adds the ability to define custom validation rules for the site config beyond what the static schema definition allows us to do
- Updates the `WebsiteStarter.config` field in Django admin so that it can accept JSON _or_ YAML. This will make it easier to use.

#### How should this be manually tested?
- Try updating the config for some `WebsiteStarter` in Django admin. If the config is invalid, you should see an error message describing the exact issue
- Try putting YAML or JSON in that field in Django admin. As long as the config is valid, it should save (as JSON)

#### Any background context you want to provide?
- As the README file says in the config_schema directory, the schema is implemented using [yamale](https://github.com/23andMe/Yamale). The specs for the schema can be found in their docs
- Our schema can be as strict as we want it to be, but there are some schema rules that I expected us to want, but couldn't find a way to support with yamale's schema language. One example rule that I implemented was the constraint that a collection item should specify only one of `folder`, `files`, or `file`. There wasn't a way to declare that via yamale that I could find. I took a look at some other schema definition libraries ([cerberus](https://docs.python-cerberus.org/en/stable/) among them) and couldn't find a way to do something like that out of the box with them either. Since (a) our schema is likely to be relatively simple, and (b) yamale's schema language is easy to understand and gives us a fair amount of what we want, I decided to go with yamale and add in an easy way to define additional rules. If we end up defining a huge amount of custom rules that get difficult to understand, we can re-evaluate and pick a different library if needed.

#### Screenshots (if appropriate)
![ss 2021-03-22 at 14 13 31 ](https://user-images.githubusercontent.com/14932219/112054410-e8c4d380-8b2b-11eb-954e-01daf8a9a9a7.png)
![ss 2021-03-22 at 17 07 58 ](https://user-images.githubusercontent.com/14932219/112058650-3a239180-8b31-11eb-848e-182984111ea1.png)
